### ISSUE TITLE
nDCG metric does not correctly truncate lists

### ISSUE THREAD

**Issue by spawliko** [lenskit/lkpy#309]
The nDCG metric has two faults whereby it fails to correctly truncate the truth list and recommendation list under certain conditions:

1. If no k is supplied, the full truth list and recommendation list are used to compute the ideal DCG and recommendation DCG, respectively. If the sizes of the input lists do not match, this causes an incorrect result. See the following example, based on "test_ndcg_wrong" in tests/test_topn_ndcg.py:
```
truth = pd.DataFrame({'item': [1, 2, 3], 'rating': [3.0, 5.0, 4.0]})
truth = truth.set_index('item')
recs = pd.DataFrame({'item': [1, 2]})

print(ndcg(recs, truth))  # incorrectly gives ~0.7344
print(ndcg(recs, truth, k=2))  # correctly gives ~0.8888
```

2. If a k is supplied that is less than the length of the recommendation list, the recommendation DCG is not computed correctly (due to "tpos" being defined before the recommendation list is truncated). See the following example, based on "test_ndcg_perfect" found in tests/test_topn_ndcg.py:
```
truth = pd.DataFrame({'item': [1, 2, 3], 'rating': [3.0, 5.0, 4.0]})
truth = truth.set_index('item')
recs = pd.DataFrame({'item': [2, 3, 1]})

print(ndcg(recs, truth))  # correctly gives 1.0
print(ndcg(recs, truth, k=2))  # incorrectly gives ~1.21
print(ndcg(recs[:2], truth, k=2))  # correctly gives 1.0
```

**Comment by mdekstrand** [comment]
2 definitely looks like a bug. Will work on a fix.

Regarding 1: when `k` is not specified, this is actually a difficult scenario, and it is not clear what the correct behavior is, because it is not clear how many results the system was _supposed_ to return. LensKit's current behavior assumes that the target list length is effectively unbounded (or at least as long as the truth frame), and thus a truncated list represents an inability to recommend further items. Another option is to assume that the list is as long as was requested, and truncate the truth; this is a quite strong assumption, and it means a recommender can obtain a higher NDCG score simply by not recommending more items.

Disambiguating this case is why `k` was introduced, and evaluations should generally always specify `k` (should make that clearer in the documentation, perhaps). From the data frame alone, the metric does not have enough information to determine a target length. Do you have any references to what assumptions are made by other implementations or authors when result lists are too short? Happy to discuss this further and be persuaded that this is a bug, but right now it looks like differing expected assumptions for an edge case.

**Comment by spawliko** [comment]
Thanks for the clarification, I now understand the utility of the current implementation and agree that 1 is not a bug.

**Comment by mdekstrand** [comment]
@spawliko thanks! I should be able to have a fix for 2 in testing later today.

**Comment by mdekstrand** [comment]
#310 fixes this for a LensKit 0.14.1 bug-fix release. Will cherry-pick that fix onto the main branch and then close the bug.
